.circleci/config.yml

```diff
@@ -22,21 +31,198 @@ anchors:
+      GRADLE_OPTS: "-Dorg.gradle.daemon=false"
       TERM: dumb
-    update_git_submodules: &update_git_submodules
-      name: Update Git Submodules
-      command: git submodule update --init --recursive
-jobs:
-  api-build-test:
-    <<: *java_defaults
+
+  # See https://circleci.com/docs/2.0/workflows/#git-tag-job-execution
+  # Job runs for master branch only
+  filter_master: &filter_master
+    filters:
+      tags:
+        ignore: /.*/
+      branches:
+        only: master
+
+  # Job runs for no branch and only for tags starting with ‘v’
+  filter_releases: &filter_releases
+    filters:
+      branches:
+        ignore: /.*/
+      tags:
+        only: /^v.*/
+
+  # Job runs for all branches EXCEPT all tag and master branches
+  filter_pr: &filter_pr
+    filters:
+      tags:
+        ignore: /.*/
+      branches:
+        ignore: master
+
+# -------------------------
+#   COMMANDS
+# Refers to https://circleci.com/docs/2.0/reusing-config/#authoring-reusable-commands
+# -------------------------
+commands:
+  activate_service_account_creds:
+    description: "Activiate CircleCI service account credential"
+    steps:
+      - run:
+          name: Activate CircleCI service account credential
+          working_directory: ~/workbench
+          # Used to call gsutil from the circle environment.
+          command: ci/activate_creds.sh api/circle-sa-key.json
+
+  git_checkout:
+    description: "git checkout and update submodules"
     steps:
       - checkout
-      - run: *update_git_submodules
-      - run: *ensure_branch_has_api_changes
+      - run:
+          name: Update git submodules
+          command: git submodule update --init --recursive
+
+  halt_job_noncode_changes:
+    description: "Halting job because no application code was changed"
+    steps:
+      - run:
+          command: |
+            if [ ${CIRCLE_BRANCH} != "" ] &&
+              [ ${CIRCLE_BRANCH} != "master" ] &&
+              ! git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep -qvE '(.md$)|(.pdf$)'; then
+                echo "No code changes in non-master branch. halting job."
+                circleci step halt
+            fi
+          name: Ensure branch has application code changes
+
+  ensure_branch_has_changes:
+    description: "Ensure branch has changes to API or UI code"
+    parameters:
+      dir_name:
+        type: enum
+        default: 'api'
+        enum: ["api", "ui"]
+    steps:
+      - run:
+          command: |
+            if [ ${CIRCLE_BRANCH} != "" ] &&
+              [ ${CIRCLE_BRANCH} != "master" ] &&
+              [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep << parameters.dir_name >>/ | wc -l | xargs) == 0 ]; then
+                echo "No relevant changes in << parameters.dir_name >> directory in non-master branch. Skipping."
+                circleci step halt
+            fi
+          name: Ensure branch has changes to API or UI code
+
+  ui_dependencies:
```

Comment:
Please name commands like functions, i.e. starting with a verb. This one could be `install_ui_dependencies` I think. You could also start with `manage` but it's less descriptive.